Engine/KDtree.h

```diff
@@ -0,0 +1,503 @@
+		int index = -2;
+    public:
+		void setIndex(int in){
+			index = in;
+		}
+		int getIndex(){
+			if(index==-2){
+
+			}
+			fprintf(stderr, "Triangle getIndex err:Triangle verticies does not set, first run setTvertices\n");
+			return index;
+		}
+
+		AABB TBoundingBox;
+		std::tuple<float, float> getBoundingEdge(int LongestAxis){
+			switch(index){
+				case -2:
+					fprintf(stderr, "Triangle getBoundingEdge err:TBoundingBox does not set, first run drawBounds\n");
+					exit(1);
+					break;
+				case -1:
+					fprintf(stderr, "Triangle getBoundingEdge err:TBoundingBox has set, but no given index\n");
+					exit(1);
+					break;
+			}
+			Vec3 temp1 = TBoundingBox.getVertexMax();
+			Vec3 temp2 = TBoundingBox.getVertexMin();
+			if(LongestAxis == AXIS_X){
+				return make_tuple(temp2.x,temp1.x);
+			}
+			else if(LongestAxis == AXIS_Y){
+				return make_tuple(temp2.y,temp1.y);
+			}
+			else if(LongestAxis == AXIS_Z){
+				return make_tuple(temp2.z,temp1.z);
+			}
+			else{
+				exit(1);
+			}
+		}
+        void setTvertices(float x1, float y1, float z1, float x2, float y2, float z2,
+                            float x3, float y3, float z3){
+			index = -1;
+            vertex[0].setVec3(x1,y1,z1);
+            vertex[1].setVec3(x2,y2,z2);
+            vertex[2].setVec3(x3,y3,z3);
+        }
+		
+		Vec3* getTverticies(){
+			if(index == -2){
+				fprintf(stderr, "Triangle getTverticies err:Triangle verticies does not set, first run setTvertices\n");
+				exit(1);
+			}
+			return vertex;
+		}
+
+		bool Intersect(const Ray& ray){
+			float o_x = ray.getOrigin().x;
+			float o_y = ray.getOrigin().y;
+			float o_z = ray.getOrigin().z;
+			Vec3 Origin(o_x,o_y,o_z);
+			float temp_x = ray.getMaxT() * ray.getDirection().x;
+			float temp_y = ray.getMaxT() * ray.getDirection().y;
+			float temp_z = ray.getMaxT() * ray.getDirection().z;
+			Vec3 ray_vector( temp_x, temp_y ,temp_z);
+			//x0 * edgeOA + x1 * edgeOB + x2 * edge OC
+			//when x0 + x1 + x2 >= 1 and x0,x1,x2 > 0 has intersect with triangle
+			//find the matrix[OA OB OC] whether singular
+			// normalize x0,x1,x2 and plus O will find the intersect point 
+			Vec3 edgeOA = vertex[0] - Origin;
+			Vec3 edgeOB = vertex[1] - Origin;
+			Vec3 edgeOC = vertex[2] - Origin;
+			//find delta of the matrix
+			float delta = Three_Vec3_delta(edgeOA,edgeOB,edgeOC);
+			float epsilon = 0.0000001;
+			if(delta < epsilon && delta > -epsilon){
+				//singular case
+				//the ray origin lies on the plane which span by AB and AC
+				return false;
+			} 
+
+			float deltax_1 = Three_Vec3_delta(ray_vector, edgeOB, edgeOC);
+			float deltax_2 = Three_Vec3_delta(edgeOA, ray_vector, edgeOC);
+			float deltax_3 = Three_Vec3_delta(edgeOA, edgeOB, ray_vector);
+			bool all_positive = (deltax_1 > 0 && deltax_2 > 0 && deltax_3 > 0);	
+			if(deltax_1 + deltax_2 + deltax_3 >= 1 && all_positive){
+				return true;
+			}
+
+			return false;
+
+ 		}
+};
+
+class Triangles{
+	public:
+		std::vector<Triangle*> tris;
```

Comment:
Who needs to take ownership of these pointers? 
Suggestion: TriangleKdtree class